BENCH, TST: benchmarks without hypothesis

tylerjereddy opened the pull request:
* Fixes gh-23344.

* The new array API `asv` benchmarks depend on `scipy.conftest`, which has a mandatory dependency on `hypothesis`. This patch makes it possible to extract the necessary array API information from `conftest` without needing the `hypothesis` dependency, so that benchmarks can run in that situation.

* I'm not sure if it is desirable long-term to have this kind of cross-talk between the `asv` benchmarks and our testing infrastructure, but this seemed simplest for now. Note that `hypothesis` is still mandatory for running our full testsuite.

tylerjereddy committed:
BENCH, TST: benchmarks without hypothesis

* Fixes gh-23344.

* The new array API `asv` benchmarks depend on `scipy.conftest`,
which has a mandatory dependency on `hypothesis`. This patch
makes it possible to extract the necessary array API information
from `conftest` without needing the `hypothesis` dependency, so
that benchmarks can run in that situation.

* I'm not sure if it is desirable long-term to have this kind
of cross-talk between the `asv` benchmarks and our testing
infrastructure, but this seemed simplest for now. Note that
`hypothesis` is still mandatory for running our full testsuite.

crusaderky reviewed (changes requested):
With this change, you can remove hypothesis from here:
https://github.com/scipy/scipy/blob/38cb4d90e871d3c8265e040c172ec818b54eb700/.circleci/config.yml#L81

tylerjereddy committed:
CI: PR 23349 revisions

* Remove `hypothesis` install from the CircleCI config, based
on reviewer feedback at:
https://github.com/scipy/scipy/pull/23349#pullrequestreview-3033920908

tylerjereddy commented:
thanks, I pushed in a revision to do that

lucascolley reviewed (approved):
thanks both!
